### PR TITLE
build-configs.yaml: add riscv to the clang configs

### PR DESCRIPTION
**Pull request by gctucker**
Add the riscv architecture to the list of configs built with clang, with all the core changes needed to be able to support it. This is used with clang-10 on mainline and clang-11 on linux-next.

### CODE REVIEW

**Comment by gctucker** [comment]
@nickdesaulniers Is this expected to work with mainline? (on `v5.10-rc2`)
```
root@2c39d1f942a7:~/kernelci-core# ./kci_build build_kernel --arch=riscv --build-env=clang-10 --kdir=linux --output=linux/build-riscv --defconfig=defconfig -j2
make KBUILD_BUILD_USER=KernelCI -Clinux -s ARCH=riscv CROSS_COMPILE=riscv64-linux-gnu- LLVM=1 CC="ccache clang" O=/root/kernelci-core/linux/build-riscv defconfig
make KBUILD_BUILD_USER=KernelCI -Clinux -j2 -s ARCH=riscv CROSS_COMPILE=riscv64-linux-gnu- LLVM=1 CC="ccache clang" O=/root/kernelci-core/linux/build-riscv
/usr/bin/as: unrecognized option '-mabi'
clang: error: assembler command failed with exit code 1 (use -v to see invocation)
make[2]: *** [../scripts/Makefile.build:283: scripts/mod/empty.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [/root/kernelci-core/linux/Makefile:1199: prepare0] Error 2
make: *** [Makefile:185: __sub-make] Error 2
```
```
root@2c39d1f942a7:~/kernelci-core# clang --version
Debian clang version 11.0.1-++20201103062553+ef4ffcafbb2-1~exp1~20201103053214.125
```

**Comment by gctucker** [comment]
(same issue on `next-20201110`)
After installing `binutils-riscv64-linux-gnu` in the Docker image, I got this:
```
root@2c39d1f942a7:~/kernelci-core# ./kci_build build_kernel --arch=riscv --build-env=clang-10 --kdir=linux --output=linux/build-riscv --defconfig=defconfig -j2
make KBUILD_BUILD_USER=KernelCI -Clinux -s ARCH=riscv CROSS_COMPILE=riscv64-linux-gnu- LLVM=1 CC="ccache clang" O=/root/kernelci-core/linux/build-riscv defconfig
make KBUILD_BUILD_USER=KernelCI -Clinux -j2 -s ARCH=riscv CROSS_COMPILE=riscv64-linux-gnu- LLVM=1 CC="ccache clang" O=/root/kernelci-core/linux/build-riscv
/tmp/empty-ef6294.s: Assembler messages:
/tmp/empty-ef6294.s:2: Error: unknown pseudo-op: `.attribute'
/tmp/empty-ef6294.s:3: Error: unknown pseudo-op: `.attribute'
clang: error: assembler command failed with exit code 1 (use -v to see invocation)
make[2]: *** [../scripts/Makefile.build:283: scripts/mod/empty.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [/root/kernelci-core/linux/Makefile:1197: prepare0] Error 2
make: *** [Makefile:185: __sub-make] Error 2
```
so I think we may be missing something else there.

**Comment by nickdesaulniers** [comment]
> Error: unknown pseudo-op: `.attribute'

That looks new to me, which config?  It looks like we are building the `defconfig`: https://github.com/ClangBuiltLinux/continuous-integration/blob/84ae7c86dad47eb5a1a367322e6ad6cd16a04a55/driver.sh#L131

Oh, we're building RISCV with `LLVM_IAS=1` which uses the assembler built into clang, rather than `riscv64-linux-gnu-as`.  This may be an interop bug between clang and gas where clang is emitting clang specific assembler attribute that gas does not understand.

**Comment by gctucker** [comment]
> > Error: unknown pseudo-op: `.attribute'
> 
> That looks new to me, which config? It looks like we are building the `defconfig`: https://github.com/ClangBuiltLinux/continuous-integration/blob/84ae7c86dad47eb5a1a367322e6ad6cd16a04a55/driver.sh#L131
> 
> Oh, we're building RISCV with `LLVM_IAS=1` which uses the assembler built into clang, rather than `riscv64-linux-gnu-as`. This may be an interop bug between clang and gas where clang is emitting clang specific assembler attribute that gas does not understand.

OK, how about using `LLVM_IAS` with all the architectures?

**Comment by gctucker** [comment]
Also that seems to get over the asm error but then I get this linker error:
```
root@52f746851b53:~/kernelci-core# export LLVM_IAS=1
root@52f746851b53:~/kernelci-core# ./kci_build build_kernel --arch=riscv --build-env=clang-10 --kdir=linux --output=linux/build-riscv --defconfig=defconfig -j1
make KBUILD_BUILD_USER=KernelCI -Clinux -s ARCH=riscv CROSS_COMPILE=riscv64-linux-gnu- LLVM=1 CC="ccache clang" O=/root/kernelci-core/linux/build-riscv defconfig
make KBUILD_BUILD_USER=KernelCI -Clinux -j1 -s ARCH=riscv CROSS_COMPILE=riscv64-linux-gnu- LLVM=1 CC="ccache clang" O=/root/kernelci-core/linux/build-riscv
clang: warning: argument unused during compilation: '-no-pie' [-Wunused-command-line-argument]
/usr/bin/ld: unrecognised emulation mode: elf64lriscv
Supported emulations: elf_x86_64 elf32_x86_64 elf_i386 elf_iamcu elf_l1om elf_k1om i386pep i386pe
clang: error: linker command failed with exit code 1 (use -v to see invocation)
llvm-objcopy: error: 'arch/riscv/kernel/vdso/vdso.so.dbg': No such file or directory
make[4]: *** [../arch/riscv/kernel/vdso/Makefile:58: arch/riscv/kernel/vdso/vdso.so] Error 1
make[3]: *** [../scripts/Makefile.build:500: arch/riscv/kernel/vdso] Error 2
make[2]: *** [../scripts/Makefile.build:500: arch/riscv/kernel] Error 2
make[1]: *** [/root/kernelci-core/linux/Makefile:1797: arch/riscv] Error 2
make: *** [Makefile:185: __sub-make] Error 2
root@52f746851b53:~/kernelci-core# clang --version
Debian clang version 11.0.1-++20201103062553+ef4ffcafbb2-1~exp1~20201103053214.125
```
FYI I'm using this Docker image interactively with kernelci-core mounted as a volume:
```
$ docker run -it -v $PWD:/root/kernelci-core kernelci/build-clang-11 /bin/bash
```

**Comment by gctucker** [comment]
And, if I manually change `/usr/bin/ld` to point to `/usr/lib/llvm-11/bin/lld` then I get to another problem:
```
  VDSOLD  arch/riscv/kernel/vdso/vdso.so.dbg
clang: warning: argument unused during compilation: '-no-pie' [-Wunused-command-line-argument]
ld: error: arch/riscv/kernel/vdso/rt_sigreturn.o:(.text+0x0): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
ld: error: arch/riscv/kernel/vdso/rt_sigreturn.o:(.text+0x2): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
ld: error: arch/riscv/kernel/vdso/rt_sigreturn.o:(.text+0x4): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
ld: error: vgettimeofday.c:(.text+0x0): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
ld: error: arch/riscv/kernel/vdso/getcpu.o:(.text+0x0): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
ld: error: arch/riscv/kernel/vdso/getcpu.o:(.text+0x2): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
ld: error: arch/riscv/kernel/vdso/getcpu.o:(.text+0x4): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
ld: error: arch/riscv/kernel/vdso/flush_icache.o:(.text+0x0): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
ld: error: arch/riscv/kernel/vdso/flush_icache.o:(.text+0x2): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
ld: error: arch/riscv/kernel/vdso/flush_icache.o:(.text+0x4): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
ld: error: arch/riscv/kernel/vdso/note.o:(.text+0x0): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
ld: error: arch/riscv/kernel/vdso/note.o:(.text+0x2): relocation R_RISCV_ALIGN requires unimplemented linker relaxation; recompile with -mno-relax
clang: error: linker command failed with exit code 1 (use -v to see invocation)
  OBJCOPY arch/riscv/kernel/vdso/vdso.so
llvm-objcopy: error: 'arch/riscv/kernel/vdso/vdso.so.dbg': No such file or directory
make[4]: *** [../arch/riscv/kernel/vdso/Makefile:58: arch/riscv/kernel/vdso/vdso.so] Error 1
make[3]: *** [../scripts/Makefile.build:500: arch/riscv/kernel/vdso] Error 2
make[2]: *** [../scripts/Makefile.build:500: arch/riscv/kernel] Error 2
make[1]: *** [/root/kernelci-core/linux/Makefile:1798: arch/riscv] Error 2
make[1]: Leaving directory '/root/kernelci-core/linux/build-riscv'
make: *** [Makefile:185: __sub-make] Error 2
make: Leaving directory '/root/kernelci-core/linux'
```
@nickdesaulniers If you've built a riscv kernel with Clang before, maybe you could share some reference instructions so we don't have to go through all that again for KernelCI?

**Comment by nickdesaulniers** [comment]
> OK, how about using LLVM_IAS with all the architectures?

I'm working on getting `LLVM_IAS=1` working on x86_64, aarch64, and arm.  aarch64 looks good to go, x86_64 is good to go with a very recent version of clang (not sure clang-11 has the last fix I'm thinking of). arm still has a few issues to be fixed in llvm to support allyesconfig.

> /usr/bin/ld: unrecognised emulation mode: elf64lriscv

The right linker isn't being found again; /usr/bin/ld is the host linker (x86) not the cross linker (riscv).

> llvm-objcopy: error: 'arch/riscv/kernel/vdso/vdso.so.dbg': No such file or directory

Ah right so this is failing to build the vdso image. There was a recent bug with this...
https://github.com/ClangBuiltLinux/linux/issues/970

Ah, it looks like we're disabling a few configs...that's...not good.
https://github.com/ClangBuiltLinux/continuous-integration/blob/84ae7c86dad47eb5a1a367322e6ad6cd16a04a55/driver.sh#L417-L428
cc @nathanchance

**Comment by nathanchance** [comment]
Yes, there are issues filed for all of those configs. It is just getting someone on the kernel or LLVM side to care :)

`ld.lld` does not work for linking a RISC-V kernel: https://github.com/ClangBuiltLinux/linux/issues/1020. `LD=riscv64-linux-gnu-ld` is needed.

You will currently run into a build failure related to `CONFIG_EFI`, which is why we disable it: https://github.com/ClangBuiltLinux/linux/issues/1143

The other configs that are disabled in our CI are ones that we blanket turn on for all architectures, the failure we see with those is https://github.com/ClangBuiltLinux/linux/issues/1144 but defconfig is not affected by that.

TL; DR:

```
$ make -skj"$(nproc)" ARCH=riscv CROSS_COMPILE=riscv64-linux-gnu- LD=riscv64-linux-gnu-ld LLVM=1 LLVM_IAS=1 defconfig

$ scripts/config -d EFI

$ make -skj"$(nproc)" ARCH=riscv CROSS_COMPILE=riscv64-linux-gnu- LD=riscv64-linux-gnu-ld LLVM=1 LLVM_IAS=1 olddefconfig all
```

_should_ work on mainline and next.

**Comment by gctucker** [comment]
OK thanks, that seems to be working with mainline but not with next.  At least I think we should have something to work with now, I'll post an update tomorrow.

**Comment by gctucker** [comment]
Erm, actually there's still an error both on mainline and next, with all the things above:
```
  AS      arch/riscv/kernel/head.o
<instantiation>:8:9: error: 2-byte data relocations not supported
 .short section_count
        ^
<instantiation>:58:8: error: expected relocatable expression
 .long (section_table - .) / 8
       ^
```

**Comment by nathanchance** [comment]
That is the error that you should get if `CONFIG_EFI` is enabled, double check that it is actually disabled. I just tested my commands on stock mainline and next and everything worked fine.

**Comment by gctucker** [comment]
Thanks, my `.config` had been reset to defconfig :facepalm:

Now it works:
```
  LD      vmlinux
  SYSMAP  System.map
  MODPOST Module.symvers
  OBJCOPY arch/riscv/boot/Image
  CC [M]  fs/nfs/flexfilelayout/nfs_layout_flexfiles.mod.o
  GZIP    arch/riscv/boot/Image.gz
  LD [M]  fs/nfs/flexfilelayout/nfs_layout_flexfiles.ko
  Kernel: arch/riscv/boot/Image.gz is ready
```
:tada: 

Next part is to get the same thing to run in KernelCI but that's just a bit more typing.

**Comment by gctucker** [comment]
The YAML config changes are breaking other arch builds as a side effect, so skipping this PR on staging for now.

**Comment by nickdesaulniers** [comment]
Thanks, as we nail down these issues, I'll make sure to circle back with KernelCI so that we can drop these workarounds.

**Comment by gctucker** [comment]
Clang builds seem to be failing with a timeout now on staging, but it works fine when I run them manually with `kci_build`.  I'll take a closer look next week.

**Comment by gctucker** [comment]
This is the list of riscv builds for linux-next now:
```
$ ./kci_build list_kernel_configs --build-config=next | grep riscv | grep clang
riscv defconfig clang-10
riscv defconfig+CONFIG_EFI=n clang-11
riscv defconfig clang-11
riscv defconfig+CONFIG_EFI=n clang-10
```

**Comment by gctucker** [comment]
Tested OK on staging:
https://staging.kernelci.org/build/id/5fb516fe0900a8d5f2b1743b/

There was just one warning:
```
clang: warning: argument unused during compilation: '-no-pie' [-Wunused-command-line-argument]
```

**Comment by gctucker** [comment]
@nickdesaulniers @nathanchance While clang-11 builds are virtually spotless, there's a problem with clang-10:
https://storage.staging.kernelci.org/next/master/next-20201118/riscv/defconfig+CONFIG_EFI=n/clang-10/build.log

If this is really a kernel/llvm issue then it's OK to enable it anyway in production.I guess it'll help you confirm when things get fixed. But let me know if there is a way to "fix" that with another config or flag and we should try to accommodate it.

**Comment by nathanchance** [comment]
Ah yes, sorry about that, I do not test RISC-V kernels with LLVM 10.0.1 for that very reason. I would recommend only enabling RISC-V builds for LLVM 11 and later.

**Comment by gctucker** [comment]
OK thanks, I've changed the YAML config accordingly to only build riscv with clang-11 on linux-next.

**Comment by broonie** [comment]
Looks good to me.
